src/rpc/blockchain.cpp

```diff
@@ -1277,26 +1277,70 @@ void writeBlockChunksAndUpdateMetadata(bool isHexEncoded, HTTPRequest &req,
                                        bool processedInBatch, const RetFormat& rf)
 {
-    auto stream = blockIndex.StreamSyncBlockFromDisk();
-    if (!stream) 
-    {
-        // Block not found on disk. This could be because we have the block
-        // header in our index but don't have the block (for example if a
-        // non-whitelisted node sends us an unrequested long chain of valid
-        // blocks, we add the headers to our index, but don't accept the block).
-        throw block_parse_error(blockIndex.GetBlockHash().GetHex() + " not found on disk");
-    }
-
     CDiskBlockMetaData metadata = blockIndex.GetDiskBlockMetaData();
     bool hasDiskBlockMetaData = !metadata.diskDataHash.IsNull();
+    
+    std::pair<bool, std::string> range_header = req.GetHeader("Range");
+    bool hasRangeHeader = range_header.first;
+
+    uint64_t offset;
+    uint64_t contentLen;
+    std::string totalLen = "*";
+    std::unique_ptr<CForwardReadonlyStream> stream;
 
     switch (rf)
     {
         case RF_BINARY:
         {
-            if (hasDiskBlockMetaData)
+            if (hasRangeHeader)
             {
-                req.WriteHeader("Content-Length", std::to_string(metadata.diskDataSize));
+                try
+                {
+                    std::string s = range_header.second;
+                    if (s.find("bytes=") != 0)
+                    {
+                        throw block_parse_error("Invalid Range header format, should starts with 'bytes='");
+                    }
+                    s.erase(0, 6);
+                    std::string delimiter = "-";
+                    std::string rs_s = s.substr(0, s.find(delimiter));
+                    s.erase(0, s.find(delimiter) + delimiter.length());
+                    std::string re_s = s;
+
+                    uint64_t rs = std::stoll(rs_s);
+                    uint64_t re = std::stoll(re_s);
+
+                    if (rs > re)
+                    {
+                        throw block_parse_error("Invalid Range parameter, start > end");
+                    }
+                    contentLen = re - rs + 1;
+
+                    if (hasDiskBlockMetaData) {
+                        if (rs >= metadata.diskDataSize)
+                        {
+                            throw block_parse_error("Invalid Range parameter, start >= data_size");
+                        }
+
+                        offset = rs;
+
+                        uint64_t remain = metadata.diskDataSize - offset;
+                        contentLen = std::min(remain, contentLen);
+                        totalLen = std::to_string(metadata.diskDataSize);
+                    }
+
+                    req.WriteHeader("Content-Length", std::to_string(contentLen));
+                    req.WriteHeader("Content-Range", "bytes " + std::to_string(offset) + "-" + 
+                        std::to_string(contentLen - 1) + "/" + totalLen);
+                }
+                catch (...) {
```

Comment:
We will catch the `block_parse_error`s thrown above here as well, and then we will loose the extra information in their exception strings when we throw again just below. Maybe it would be better to also add here:
````
catch(const block_parse_error&)
{
    throw;
}
````